[#3288] Migrate flatpages to CMS

pi-sigma opened the pull request:
- [x] add CMS plugin and pages for displaying content of flatpages
- [x] add data migration to move content of flatpages to CMS pages 
- [x] update admin index fixture (remove flatpages)
- [x] update sitemap (replace refs to flatpages with refs to cms pages)

Taiga: https://taiga.maykinmedia.nl/project/open-inwoner/us/3288

pi-sigma committed:
feat: add CMS plugin and model for displaying content of flatpages

pi-sigma committed:
chore: migrate flatpages content to CMS

pi-sigma committed:
chore: update admin index fixture (remove flatpages)

pi-sigma committed:
refactor: update sitemap to use CMS pages instead of flatpages

swrichards commented on src/open_inwoner/cms/footer/migrations/0002_migrate_flatpages_content_to_cms.py:
@pi-sigma a note that adding the CMS pages will trigger the django elasticsearch handler to immediately index these pages, which means the migration will fail if elasticsearch is unavailable. We should look into disable the relevant signal in the `create_cms_pages` handlers.

swrichards commented on src/open_inwoner/conf/fixtures/django-admin-index.json:
```diff
@@ -201,7 +201,9 @@
+        "translations": {
+            "nl": "Bestand"
+        },
         "name": "Configuratie",
```
This looks a bit strange to me, do we need it? (Bestand is now the name of the menu root)

pi-sigma replied:
Artifact from developing/debugging (I had to put something in there to save the application groups). Removed it.

swrichards commented:
@pi-sigma there was a question during the sprint review regarding the contactformulier page, whether the surrounding text there is also editable. I wasn't sure whether this related to this PR or is a separate issue, did you look into this at all?

swrichards commented on src/open_inwoner/cms/footer/migrations/0002_migrate_flatpages_content_to_cms.py:
```diff
@@ -0,0 +1,92 @@
+
+from django.db import migrations
+
+# from django.contrib.flatpages.models import FlatPage
```
```suggestion
```

swrichards commented on src/open_inwoner/cms/footer/migrations/0002_migrate_flatpages_content_to_cms.py:
```diff
@@ -0,0 +1,92 @@
+from ..cms_plugins import CMSFlatPagePlugin
+
+
+def create_cms_pages(apps, schema_editor):
```
Were you able to check for disabling the signals for auto ES indexing here, or was that not feasible?

pi-sigma replied:
Which signals do you mean? We have a task for rebuilding search the index on a regular basis; I wasn't aware of signals that trigger rebuilding.

swrichards replied:
Since we added CMS pages to the search index, adding nen CMS pages will trigger one of these:

https://django-elasticsearch-dsl.readthedocs.io/en/latest/es_index.html#signals

The idea is to automatically keep the indexes in sync, but mostly this seems to be a nuisance (we use the Celery job for that).

swrichards replied:
Scratch that, this is for post-index, let me double check what happened on my machine and if I can reproduce.

swrichards commented on src/open_inwoner/configurations/models.py:
```diff
@@ -664,8 +663,11 @@ def get_accent_color(self):
     @property
-    def get_ordered_flatpages(self):
-        return self.flatpages.order_by("ordered_flatpages")
+    def get_ordered_cms_pages(self):
```
```suggestion
    def ordered_cms_pages(self):
```
The imperative verb suggests it should be a method: I would either rename or remove `@property` (I am aware this pattern was introduced earlier, but let's try and break it :).

pi-sigma replied:
Was tempted to do this but shied away for fear of making things more confusing. If we're on the same page, though, let's go for it.

swrichards commented on src/open_inwoner/cms/footer/cms_plugins.py:
```diff
@@ -1,21 +1,69 @@
+
+    def render(self, context, instance, placeholder):
+        cms_plugin_placeholder = instance.placeholder
+        cms_page = Page.objects.filter(placeholders=cms_plugin_placeholder).first()
```
Should we log an error if there's multiple objects?

pi-sigma replied:
I added a warning. Shouldn't happen but...

swrichards commented on src/open_inwoner/cms/footer/migrations/0002_migrate_flatpages_content_to_cms.py:
```diff
@@ -0,0 +1,92 @@
+    FlatPage = apps.get_model("flatpages", "FlatPage")
+
+    for flatpage in FlatPage.objects.all():
+        flatpage.delete()
```
```suggestion
    FlatPage.objects.all().delete()
```

swrichards commented on src/open_inwoner/cms/footer/models.py:
```diff
@@ -0,0 +1,18 @@
+        blank=True,
+        help_text=_("The title of the page"),
+    )
+    content = models.TextField(
```
To be clear, this is rich content?

pi-sigma replied:
Yes but it's considered safe because only exposed to staff users.

pi-sigma committed:
fix: disable elasticsearch signals for CMS page document

    - In preparation for the migration of flatpage content to
      CMS pages, we disable the signals for the CMS page document
      to avoid automatic re-building of the search index.
